frame/support/procedural/src/pallet/parse/call.rs

```diff
@@ -0,0 +1,237 @@
+
+/// List of additional token to be used for parsing.
+mod keyword {
+	syn::custom_keyword!(DispatchResultWithPostInfo);
+	syn::custom_keyword!(Call);
+	syn::custom_keyword!(OriginFor);
+	syn::custom_keyword!(weight);
+	syn::custom_keyword!(compact);
+	syn::custom_keyword!(T);
+	syn::custom_keyword!(pallet);
+}
+
+/// Definition of dispatchables typically `impl<T: Config> Pallet<T> { ... }`
+pub struct CallDef {
+	/// The where_clause used.
+	pub where_clause: Option<syn::WhereClause>,
+	/// A set of usage of instance, must be check for consistency with trait.
+	pub instances: Vec<helper::InstanceUsage>,
+	/// The index of call item in pallet module.
+	pub index: usize,
+	/// Information on methods (used for expansion).
+	pub methods: Vec<CallVariantDef>,
+	/// The span of the attribute.
+	pub attr_span: proc_macro2::Span,
+}
+
+/// Definition of dispatchable typically: `#[weight...] fn foo(origin .., param1: ...) -> ..`
+pub struct CallVariantDef {
+	/// Function name.
+	pub fn_: syn::Ident,
+	/// Information on args: `(is_compact, name, type)`
+	pub args: Vec<(bool, syn::Ident, Box<syn::Type>)>,
+	/// Weight formula.
+	pub weight: syn::Expr,
+	/// Docs, used for metadata.
+	pub docs: Vec<syn::Lit>,
+}
+
+/// Attributes for functions in call impl block.
+/// Parse for `#[pallet::weight = expr]`
+pub struct FunctionAttr {
+	weight: syn::Expr,
+}
+
+impl syn::parse::Parse for FunctionAttr {
+	fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+		input.parse::<syn::Token![#]>()?;
+		let content;
+		syn::bracketed!(content in input);
+		content.parse::<keyword::pallet>()?;
+		content.parse::<syn::Token![::]>()?;
+
+		content.parse::<keyword::weight>()?;
+
+		let weight_content;
+		syn::parenthesized!(weight_content in content);
+		Ok(FunctionAttr {
+			weight: weight_content.parse::<syn::Expr>()?,
+		})
+	}
+}
+
+/// Attribute for arguments in function in call impl block.
+/// Parse for `#[pallet::compact]|
+pub struct ArgAttrIsCompact;
+
+impl syn::parse::Parse for ArgAttrIsCompact {
+	fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+		input.parse::<syn::Token![#]>()?;
+		let content;
+		syn::bracketed!(content in input);
+		content.parse::<keyword::pallet>()?;
+		content.parse::<syn::Token![::]>()?;
+
+		content.parse::<keyword::compact>()?;
+		Ok(ArgAttrIsCompact)
+	}
+}
+
+/// Check the syntax is `OriginFor<T>`
+pub fn check_dispatchable_first_arg_type(ty: &syn::Type) -> syn::Result<()> {
+	let expected = "expect `OriginFor<T>`";
+
+	pub struct CheckDispatchableFirstArg;
+	impl syn::parse::Parse for CheckDispatchableFirstArg {
+		fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+			input.parse::<keyword::OriginFor>()?;
+			input.parse::<syn::Token![<]>()?;
+			input.parse::<keyword::T>()?;
+			input.parse::<syn::Token![>]>()?;
+
+			Ok(Self)
+		}
+	}
+
+	syn::parse2::<CheckDispatchableFirstArg>(ty.to_token_stream())
+		.map_err(|e| {
+			let msg = format!("Invalid type: {}", expected);
```

Comment:
I don't get the existential purpose of `expected`.

Reply:
I mean, it is a variable in a tiny function use just once.. I would just get rid of the `format!`.

Reply:
yes I agree, done in https://github.com/paritytech/substrate/pull/6877/commits/0adbb30495970b7022c8dfa701c78080c7c1c875